File: web/lib/sport.ts
/**
 * Sport-type constants shared across the app.
 *
 * Mirrors `src/domain/dataset/sport.py` — kept manually in sync rather than
 * generated, since it's a short, rarely-changed list. Four families: running
 * (what the app was built for), cycling (added alongside it), hiking and
 * swimming. Home's volume totals and PR ladder stay running-only, and the
 * Analysis section's panels can't mix families (see that Python module's
 * docstring for why: GAP and modelled power are running biomechanics, not
 * comparable to a ride, a hike or a swim) — but Home's "latest activity" has
 * no such comparability problem and shows any sport.
 */

export const RUNNING_SPORT_TYPES = ["Run", "TrailRun", "VirtualRun"];
export const CYCLING_SPORT_TYPES = [
  "Ride", "MountainBikeRide", "GravelRide", "VirtualRide",
];
export const HIKING_SPORT_TYPES = ["Hike", "Walk"];
export const SWIMMING_SPORT_TYPES = ["Swim"];

/** Colour a sport type reads as, wherever one is shown: the calendar's
 * completed-session chip, the session detail's sport tag, and a training
 * week's totals share this so a run — trail, road, or virtual — looks the
 * same everywhere it appears. One tone per family, not per exact sport
 * type: the four families are the unit a rider or a hiker thinks in, not
 * the distinction between a trail and a road run. */
export type SportTone = "running" | "hiking" | "cycling" | "swimming" | "neutral";

export function sportTone(sportType: string): SportTone {
  if (RUNNING_SPORT_TYPES.includes(sportType)) return "running";
  if (CYCLING_SPORT_TYPES.includes(sportType)) return "cycling";
  if (HIKING_SPORT_TYPES.includes(sportType)) return "hiking";
  if (SWIMMING_SPORT_TYPES.includes(sportType)) return "swimming";
  return "neutral";
}

File: src/domain/dataset/sport.py
"""Which broad sport family an activity belongs to.

Four families exist: **running** (the app's original domain — "Run", "TrailRun",
"VirtualRun"), **cycling** (every one of Strava's bike ``sport_type`` values —
a gravel or mountain ride is still "usual cycling" to every panel here, since
none of them distinguish bike sub-types), **hiking** ("Hike", "Walk"), and
**swimming** ("Swim"). A panel's data has to stay within one family — GAP and
the modelled power-per-kg are running biomechanics, calibrated to a "balanced
runner" reference curve that means nothing for a bike, a hike or a swim, and
even a mechanically sport-agnostic number like "fastest 10 km" is not a
comparable figure between a foot split, a bike split, a hiking split and a
swim split. See :mod:`src.domain.dataset.features` for where running-only
columns are guarded, and :mod:`src.usecases.resolve_panel_data` for where a
mixed selection is resolved down to one family.
"""

RUNNING_SPORT_TYPES = frozenset({"Run", "TrailRun", "VirtualRun"})
CYCLING_SPORT_TYPES = frozenset({
    "Ride", "MountainBikeRide", "GravelRide", "VirtualRide"
})
HIKING_SPORT_TYPES = frozenset({"Hike", "Walk"})
SWIMMING_SPORT_TYPES = frozenset({"Swim"})

RUNNING = "running"
CYCLING = "cycling"
HIKING = "hiking"
SWIMMING = "swimming"


def sport_family(sport_type: str) -> str:
    """``"cycling"``/``"hiking"``/``"swimming"`` for a recognised ride/hike/swim,
    ``"running"`` for everything else.

    Running is the fallback rather than "unknown" because it is what every
    activity in this app was before the other families existed — an
    oddly-typed or summary-only row should keep behaving exactly as it did.
    """
    if sport_type in CYCLING_SPORT_TYPES:
        return CYCLING
    if sport_type in HIKING_SPORT_TYPES:
        return HIKING
    if sport_type in SWIMMING_SPORT_TYPES:
        return SWIMMING
    return RUNNING
